Extend solidity-frontend to support Method Overload, Inheritance and more return types for functions calls.

Sleepytt1210 opened the pull request:
- Unique id is create for methods' symbol to avoid name clashes. This allows method overloading to be correctly converted.
The symbol.id of the methods are formalised as `id = “c:@C@” + contract_name + “@F@” + function_name + “#” + function_id`.

- For method calling, the function declaration node is retrieved by using the *"referencedDeclaration"* field which returned the id of the original function declaration node. 

- Added support for more return types by using get_type_descriptions to parse the *"typeDescriptions"* fields.

- Also fixed a bug where integer literals were always returned as bit vector type of `width: 32` for ReturnStatement.

Relevant test cases are added in `regression/esbmc-solidity/` to test the correctness.

Sleepytt1210 committed:
Fix typo

Sleepytt1210 committed:
Adjust  methods symbol id to include contract name

Use function parameters for return statement.

merge 2

Sleepytt1210 committed:
Fix return type and add  function overload support

Add polymorphism test cases

Add different return type test cases

Fix MemberCall to get correct  referred contract name.

Change description

Sleepytt1210 committed:
Merge duplicate  call

Added comment

Reformat description

Sleepytt1210 committed:
Add polymorphism inheritance tests

Sleepytt1210 committed:
Add return parameters  array type support

Change comment

Rename test cases.

Fix test cases descriptions.

Fix description

kunjsong01 commented on regression/esbmc-solidity/return_4/contract.sol:
```diff
@@ -0,0 +1,43 @@
+pragma solidity >=0.5.0;
+
+// Ensure functions can return different type of value.
+contract A {
```
What's the purpose of this test case?

Sleepytt1210 replied:
My mistake, this should test for member calls of other integer return types. I will add the member calls assertion below.

kunjsong01 commented on regression/esbmc-solidity/polymorphism_1/polymorphism.sol:
```diff
@@ -0,0 +1,26 @@
+
+contract C {
+    function test_poly() public {
+        A contract_A = new A();
```
This test case(TC) doesn't cover polymorphism. Please update the name for this TC.

kunjsong01 replied:
also for other TCs named `polymorphism_*`

Sleepytt1210 replied:
Will do.

kunjsong01 commented on regression/esbmc-solidity/polymorphism_4/polymorphism.sol:
```diff
@@ -0,0 +1,23 @@
+// while contract_A.poly(300) should call poly(uint16).
+contract A {
+
+    function poly(uint16 b) public pure returns (uint16) {
```
This TC tests function overloading. Please update the TC name.

kunjsong01 commented on src/solidity-frontend/solidity_convert.h:
```diff
@@ -186,6 +186,248 @@ class solidity_convertert
     SolidityGrammar::ElementaryTypeNameT &type,
     typet &out);
+  
+  /*
```
There's no implementation for the method declarations below. please remove if not needed.

Sleepytt1210 replied:
No problem.

Sleepytt1210 replied:
Hi, I think the implementation can be found at solidity_convert.cpp:2074. It's used in the get_elementary_type() function. So I'll keep it.

ChenfengWei0 replied:
I think he means the code changes in lines 189 to 430.

Sleepytt1210 replied:
Noted, thanks!

kunjsong01 reviewed (changes requested):
1. This PR adds support for basic inheritance rather than polymorphism. Please update the PR title/description, as well as the names for the test cases added. 
2. For a passing TC, we usually add failing TC (see test cases in `src/regression/esbmc` - `to_union_0` and `to_union_0_fail`)

kunjsong01 commented on regression/esbmc-solidity/polymorphism_4/polymorphism.sol:
```diff
@@ -0,0 +1,23 @@
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity >=0.5.0;
+
+// Methods overloadin: contract_A.poly(-2) should call the poly(int8) 
```
typo - overloading

ChenfengWei0 commented on regression/esbmc-solidity/polymorphism_1/polymorphism.sol:
```diff
@@ -0,0 +1,26 @@
+    function test_poly() public {
+        A contract_A = new A();
+        B contract_B = new B();
+        assert(contract_A.poly() == 21);
```
May I ask if the verification fails if we change this assertion to `assert(contract_A.poly() == 42);`

Sleepytt1210 replied:
Yes, the verification failed with counterexample given:
```
State 1 file polymorphism.sol line 23 function test_poly thread 0
----------------------------------------------------
Violated property:
  file polymorphism.sol line 23 function test_poly
  assertion
  (signed int)return_value$_poly$3 == 42


VERIFICATION FAILED
```

ChenfengWei0 commented:
Thank you very much for your work. Here is my comments:
-  You might need to consider the implicit constructor. 
    - In current ESBMC-Solidity, if there is no constructor function, we will manually add a default constructor by `add_implicit_constructor()`. 
- So you might need to change the `id` format in:
```c++
 bool solidity_convertert::add_implicit_constructor()
{
  std::string name, id;
  name = id = current_contractName;     // set the idstring format 
```
and in any place where a function symbol is looked up by calling `context.find_symbol(id)`, e.g
```c++
bool solidity_convertert::get_implicit_ctor_call(
  const int ref_decl_id,
  exprt &new_expr)
{
  std::string name, id;
  if(get_contract_name(ref_decl_id, name))
    return true;
  id = name;      // set the id format 
  if(context.find_symbol(id) == nullptr)
    return true;
```

ChenfengWei0 commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -1179,12 +1186,24 @@ bool solidity_convertert::get_expr(
     std::string name, id;
     name = callee_expr_json["memberName"];
-    id = name;
+    id = "c:@C@" + ref_contract_name + "@F@" + name + "#" + i2string(caller_id);
```
Is it possible to obtain the `id` uniformly by `get_function_definition_name()`?

Sleepytt1210 replied:
Yes, I will make corresponding changes.

ChenfengWei0 commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -1179,12 +1186,24 @@ bool solidity_convertert::get_expr(
   case SolidityGrammar::ExpressionT::MemberCallClass:
   {
     assert(expr.contains("expression"));
+    expr.dump(4, ' ', false, nlohmann::detail::error_handler_t::strict).c_str();
```
May I ask if you can add some comments here?

Sleepytt1210 replied:
Sorry about this, I think I need to remove this. I wrote this to bypass the inline function call error during the gdb debugging session.

ChenfengWei0 commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -471,7 +476,8 @@ bool solidity_convertert::get_access_from_decl(
 bool solidity_convertert::get_function_definition(
-  const nlohmann::json &ast_node)
+  const nlohmann::json &ast_node,
+  exprt &new_expr)
```
Why do we need the `new_expr` here?

Sleepytt1210 replied:
It was a dummy expression, I will remove it.

lucasccordeiro commented:
Can you add this case as a regression test?Sent from my iPhoneOn 12 May 2023, at 22:25, Sleepytt1210 ***@***.***> wrote:﻿
@Sleepytt1210 commented on this pull request.



In regression/esbmc-solidity/polymorphism_1/polymorphism.sol:
> +    function poly() public pure returns (uint8) {
+        return 21;
+    }
+}
+
+contract B {
+    function poly() public pure returns (uint8) {
+        return 42;
+    }
+}
+
+contract C {
+    function test_poly() public {
+        A contract_A = new A();
+        B contract_B = new B();
+        assert(contract_A.poly() == 21);

Yes, the verification failed with counterexample given:
State 1 file polymorphism.sol line 23 function test_poly thread 0
----------------------------------------------------
Violated property:
  file polymorphism.sol line 23 function test_poly
  assertion
  (signed int)return_value$_poly$3 == 42


VERIFICATION FAILED


—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>

Sleepytt1210 committed:
Function ID uniformly obtainable.

Remove unused statement.

Removed unused parameter.

Sleepytt1210 committed:
Fix literal return type always returning 32 bits.

Return type success test cases.

Sleepytt1210 committed:
Add missing file

Added fail test cases for function overload

Added inheritance fail test cases

Sleepytt1210 committed:
Removes unimplemented declarations.

Fix warning

ChenfengWei0 commented:
> Thank you very much for your work. Here is my comments:
> 
>     - You might need to consider the implicit constructor.
>       
>       - In current ESBMC-Solidity, if there is no constructor function, we will manually add a default constructor by `add_implicit_constructor()`.
> 
>     - So you might need to change the `id` format in:
> 
> ```c++
>  bool solidity_convertert::add_implicit_constructor()
> {
>   std::string name, id;
>   name = id = current_contractName;     // set the idstring format 
> ```
> 
> and in any place where a function symbol is looked up by calling `context.find_symbol(id)`, e.g
> 
> ```c++
> bool solidity_convertert::get_implicit_ctor_call(
>   const int ref_decl_id,
>   exprt &new_expr)
> {
>   std::string name, id;
>   if(get_contract_name(ref_decl_id, name))
>     return true;
>   id = name;      // set the id format 
>   if(context.find_symbol(id) == nullptr)
>     return true;
> ```

---

May I ask if you have a plan to implement this? It could be another PR though. 

Also, you may want to resolve the build error. The PR can only be merged if all checks are passed. 
```sh
In file included from ../src/solidity-frontend/solidity_convert.cpp:1:
../src/solidity-frontend/solidity_convert.h: In constructor 'solidity_convertert::solidity_convertert(contextt&, nlohmann::json&, const string&, const string&)':
../src/solidity-frontend/solidity_convert.h:48:15: error: 'solidity_convertert::current_contractName' will be initialized after [-Werror=reorder]
   48 |   std::string current_contractName;
      |               ^~~~~~~~~~~~~~~~~~~~
In file included from ../src/solidity-frontend/solidity_convert.cpp:1:
../src/solidity-frontend/solidity_convert.h:38:16: error:   'unsigned int solidity_convertert::current_scope_var_num' [-Werror=reorder]
   38 |   unsigned int current_scope_var_num;
      |                ^~~~~~~~~~~~~~~~~~~~~
../src/solidity-frontend/solidity_convert.cpp:15:1: error:   when initialized here [-Werror=reorder]
   15 | solidity_convertert::solidity_convertert(
      | ^~~~~~~~~~~~~~~~~~~
cc1plus: all warnings being treated as errors
```

Sleepytt1210 commented:
> > Thank you very much for your work. Here is my comments:
> > ```
> > - You might need to consider the implicit constructor.
> >   
> >   - In current ESBMC-Solidity, if there is no constructor function, we will manually add a default constructor by `add_implicit_constructor()`.
> > 
> > - So you might need to change the `id` format in:
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```c++
> >  bool solidity_convertert::add_implicit_constructor()
> > {
> >   std::string name, id;
> >   name = id = current_contractName;     // set the idstring format 
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and in any place where a function symbol is looked up by calling `context.find_symbol(id)`, e.g
> > ```c++
> > bool solidity_convertert::get_implicit_ctor_call(
> >   const int ref_decl_id,
> >   exprt &new_expr)
> > {
> >   std::string name, id;
> >   if(get_contract_name(ref_decl_id, name))
> >     return true;
> >   id = name;      // set the id format 
> >   if(context.find_symbol(id) == nullptr)
> >     return true;
> > ```
> 
> May I ask if you have a plan to implement this? It could be another PR though.
> 
> Also, you may want to resolve the build error. The PR can only be merged if all checks are passed.
> 
> ```shell
> In file included from ../src/solidity-frontend/solidity_convert.cpp:1:
> ../src/solidity-frontend/solidity_convert.h: In constructor 'solidity_convertert::solidity_convertert(contextt&, nlohmann::json&, const string&, const string&)':
> ../src/solidity-frontend/solidity_convert.h:48:15: error: 'solidity_convertert::current_contractName' will be initialized after [-Werror=reorder]
>    48 |   std::string current_contractName;
>       |               ^~~~~~~~~~~~~~~~~~~~
> In file included from ../src/solidity-frontend/solidity_convert.cpp:1:
> ../src/solidity-frontend/solidity_convert.h:38:16: error:   'unsigned int solidity_convertert::current_scope_var_num' [-Werror=reorder]
>    38 |   unsigned int current_scope_var_num;
>       |                ^~~~~~~~~~~~~~~~~~~~~
> ../src/solidity-frontend/solidity_convert.cpp:15:1: error:   when initialized here [-Werror=reorder]
>    15 | solidity_convertert::solidity_convertert(
>       | ^~~~~~~~~~~~~~~~~~~
> cc1plus: all warnings being treated as errors
> ```

Thank you for informing that! I will resolve the errors for this PR, but the implicit constructor may be implemented in another PR because I am having exams soon.

Sleepytt1210 committed:
Fix pointer cast expr bug.

Fix tests.

Set inheritance_4 to a KNOWNBUG

Fix warning

Set result to FAILED.

Fix descriptions

Sleepytt1210 commented:
> > Thank you very much for your work. Here is my comments:
> > ```
> > - You might need to consider the implicit constructor.
> >   
> >   - In current ESBMC-Solidity, if there is no constructor function, we will manually add a default constructor by `add_implicit_constructor()`.
> > 
> > - So you might need to change the `id` format in:
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```c++
> >  bool solidity_convertert::add_implicit_constructor()
> > {
> >   std::string name, id;
> >   name = id = current_contractName;     // set the idstring format 
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and in any place where a function symbol is looked up by calling `context.find_symbol(id)`, e.g
> > ```c++
> > bool solidity_convertert::get_implicit_ctor_call(
> >   const int ref_decl_id,
> >   exprt &new_expr)
> > {
> >   std::string name, id;
> >   if(get_contract_name(ref_decl_id, name))
> >     return true;
> >   id = name;      // set the id format 
> >   if(context.find_symbol(id) == nullptr)
> >     return true;
> > ```
> 
> May I ask if you have a plan to implement this? It could be another PR though.
> 
> Also, you may want to resolve the build error. The PR can only be merged if all checks are passed.
> 
> ```shell
> In file included from ../src/solidity-frontend/solidity_convert.cpp:1:
> ../src/solidity-frontend/solidity_convert.h: In constructor 'solidity_convertert::solidity_convertert(contextt&, nlohmann::json&, const string&, const string&)':
> ../src/solidity-frontend/solidity_convert.h:48:15: error: 'solidity_convertert::current_contractName' will be initialized after [-Werror=reorder]
>    48 |   std::string current_contractName;
>       |               ^~~~~~~~~~~~~~~~~~~~
> In file included from ../src/solidity-frontend/solidity_convert.cpp:1:
> ../src/solidity-frontend/solidity_convert.h:38:16: error:   'unsigned int solidity_convertert::current_scope_var_num' [-Werror=reorder]
>    38 |   unsigned int current_scope_var_num;
>       |                ^~~~~~~~~~~~~~~~~~~~~
> ../src/solidity-frontend/solidity_convert.cpp:15:1: error:   when initialized here [-Werror=reorder]
>    15 | solidity_convertert::solidity_convertert(
>       | ^~~~~~~~~~~~~~~~~~~
> cc1plus: all warnings being treated as errors
> ```

Hi,

I have a test that will not be fixed in this PR. Is there any way to mark it be ignored, or should I just remove the test case?

lucasccordeiro commented:
You could use `KNOWNBUG` instead of `CORE`.

ChenfengWei0 commented:
@kunjsong01 Do you know how to resolve the C/C++ code-style error? 👀

Sleepytt1210 commented:
> @kunjsong01 Do you know how to resolve the C/C++ code-style error? 👀

I think I will have to code it like the git diff suggested. I will fix this asap.

Sleepytt1210 committed:
Return correct format for make_pointee_type

Sleepytt1210 committed:
Clang-format

lucasccordeiro commented:
@kunjsong01 and @ChenfengWei0 : are you happy with this PR?

kunjsong01 reviewed (approved):
LGTM

lucasccordeiro commented:
@Sleepytt1210 : could you please rebase this PR with the master branch? You could also squash some commits with the command.

ChenfengWei0 reviewed (approved):
LGTM. Many thanks @Sleepytt1210

Sleepytt1210 commented:
> @Sleepytt1210 : could you please rebase this PR with the master branch? You could also squash some commits with the command.

Yes, will do.

lucasccordeiro commented:
Many thanks, @Sleepytt1210. As soon as you rebase and squash your commits, we'll merge your code into the master.

lucasccordeiro commented:
@Sleepytt1210:

````
The following tests FAILED:
	832 - regression/esbmc-solidity/constructor_1 (Failed)
	834 - regression/esbmc-solidity/constructor_3 (Failed)
	835 - regression/esbmc-solidity/constructor_4 (Failed)
	836 - regression/esbmc-solidity/constructor_5 (Failed)
	837 - regression/esbmc-solidity/constructor_6 (Failed)
	881 - regression/esbmc-solidity/op_ternary_2 (Failed)
	885 - regression/esbmc-solidity/return_1 (Failed)
	886 - regression/esbmc-solidity/return_2 (Failed)
	887 - regression/esbmc-solidity/return_3 (Failed)
Errors while running CTest
````

Sleepytt1210 commented:
> @Sleepytt1210:
> 
> ```
> The following tests FAILED:
> 	832 - regression/esbmc-solidity/constructor_1 (Failed)
> 	834 - regression/esbmc-solidity/constructor_3 (Failed)
> 	835 - regression/esbmc-solidity/constructor_4 (Failed)
> 	836 - regression/esbmc-solidity/constructor_5 (Failed)
> 	837 - regression/esbmc-solidity/constructor_6 (Failed)
> 	881 - regression/esbmc-solidity/op_ternary_2 (Failed)
> 	885 - regression/esbmc-solidity/return_1 (Failed)
> 	886 - regression/esbmc-solidity/return_2 (Failed)
> 	887 - regression/esbmc-solidity/return_3 (Failed)
> Errors while running CTest
> ```

I think I made some errors during rebase, I will fix this.

Sleepytt1210 committed:
[solidity] Clang-format

lucasccordeiro commented:
Many thanks for this PR, @Sleepytt1210.

lucasccordeiro commented:
Many thanks for this PR, @Sleepytt1210.

Sleepytt1210 commented:
You're welcome, thank you for the opportunity too.
